### ISSUE TITLE
one little confusion about the loss_fn_kd function

### ISSUE THREAD

**Issue by HLBayes** [GMvandeVen/continual-learning#18]
Many thanks for your impressive project. Here I am a few confused about the `.detach()` in the below code,
https://github.com/GMvandeVen/continual-learning/blob/a02db26d3b10754abdc4a549bdcde6af488c94e0/utils.py#L35

which is defined in https://github.com/GMvandeVen/continual-learning/blob/a02db26d3b10754abdc4a549bdcde6af488c94e0/utils.py#L18

Refer to the blog, [PyTorch .detach() method ](http://www.bnikolic.co.uk/blog/pytorch-detach.html), `.detach()` will take the `targets_norm` as one fixed parameter in the the `KD_loss`, and the  backpropagation will not update the parameters along the `targets_norm` related branch.

However, in your another project, [brain-inspired-replay](https://github.com/GMvandeVen/brain-inspired-replay), the same loss function, [`loss_fn_kd`](https://github.com/GMvandeVen/brain-inspired-replay/blob/1a030f75666c656416e1ca02466758ca32cf2fe4/models/utils/loss_functions.py#L12) uses,

```python
 targets_norm = torch.cat([targets_norm, zeros_to_add], dim=1)
```
as shown in line 29, in which no `.detach()` is attached.

 Although the same results all these two types I have tested, I am still confused about how does the second type work?

**Comment by GMvandeVen** [comment]
Hi @HLBayes, I'm sorry for the very late reply!
This difference between the two repositories is indeed confusing. In both repositories, the `.detach()` operation is actually not needed. As you indicate in your comment, the `.detach()` operation stops backpropagation as it resets any gradients being tracked. However, in both repositories, the `target_norm` variable already did not have any gradients being tracked, as that variable was computed using `with torch.no_grad():` (as for example here: [link](https://github.com/GMvandeVen/continual-learning/blob/a02db26d3b10754abdc4a549bdcde6af488c94e0/train.py#L189)). Sorry for the confusion!
